x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanStreamInput.java

```diff
@@ -69,6 +69,8 @@ public NameId apply(long streamNameId) {
 
     private EsField[] esFieldsCache = new EsField[64];
 
+    private String[] stringCache = new String[64];
```

Comment:
How about using a list or a pool of arrays? Growing the array requires double the amount of memory even if we want just a few items.

Reply:
Separately, make sure to clean this array/list and the one above it in the close() method.

Reply:
> How about using a list or a pool of arrays?

I gave priority to efficiency and simplicity here.  

The maximum amount of wasted memory with a single array is a few tens of kb heap for a query with ten thousand fields, that seems pretty reasonable; I'm not sure a space optimization would justify the complexity and the computational cost of managing and accessing an array pool or a list.

Reply:
It's the memory trashing that I'm concerned about but it might be a minute one - @original-brownbear do we have a String/object pool anywhere that we could reuse?

Reply:
@costin we have org.elasticsearch.common.util.StringLiteralDeduplicator but that is also doing interning and is thread-safe. Seems you don't need either so it's probably needlessly heavy for this use case?

Reply:
@original-brownbear Thanks for the quick reply - right, no point for interning. 
We're doing a lot of string ser/de which seems a common enough and I'm wondering whether this hasn't been addressed before and what reusable options there are in ES.

Reply:
@costin this is definitely a common problem that we have across a bunch of different pieces of functionality. Sadly, we don't have a common solution to it. It's hard to have a one size fits all solution here really.
If you're only serialising to on-heap buffers here, I'd suggest just building the symbol table as you serialise and then prepend the table as a list keyed by id after you serialised the full thing. That's the approach I took for some parts of field_caps and assuming you never actually "stream" with this thing, it's clearly the most performant option in theory, especially on the read side where you save all the resizing of a list/array/map of symbols.

Reply:
do you have an example of prepending? We're doing the writing on the stream so there's an assumption that the data might be streamed (unclear whether this will happen or not in the future).

Reply:
@costin you have 2 options here :)

You could implement something brute force like `org.elasticsearch.common.io.stream.RecyclerBytesStreamOutput#writeWithSizePrefix` where you serialize all but the symbol table to a `RecyclerBytesStreamOutput` and then write the symbol table you build while you did that first and then write the rest of the message back to the stream. That's not the absolutely most performant thing to do but might be good enough :)
Otherwise what I did in field caps is just compute the lookup table by other means (not as part of the serialization) first and then run the serialization in another loop but that also has it's cost naturally. If this is really performance critical I could push on https://github.com/elastic/elasticsearch/pull/110557 again which would make the serializing in two chunks bit effectively free?